### PR TITLE
Better rolling

### LINKED ISSUE #363: Rolling update will not complete if CC is restarted in the middle
When CC crashes or is restarted in the middle of Rolling Update, it will never finish it after restart. We need to track rolling updates and be able to complete them in such situations.

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Enhancement / new feature

### Description

This PR fixes #363 by changing the way StatefulSets are updated/rolled so that if the CO crashes part way through it will complete the roll when it restarts.

Using the same logic as before, whenever we patch in a way that needs a rolling update we update the SS template, incrementing the value of a 'generation' annotation. This happens as part of the SS patch. Previously, we only did the rolling update if the patch operation indicated that it was needed, but now (because a rolling update may be necessary even if the last patch was a noop) we always call `maybeRollingUpdate()`, and let that method decide which pods to roll. Since rolled pods will pick up the new generation number from the SS's template we can distinguish rolled pods from unrolled ones. Thus we only actually rolls pods whose generation annotation does not match the StatefulSets, and so `maybeRollingUpdate()` will be a noop iff all the pods' generation numbers match.

The changes are very noisy because we no longer need to return a boolean from the reconcile method on the SS: We actually return the whole updated SS, wrapped in a `ReconciliationResult` (it would be enough to return the generation number), which means we no longer need the type parameter `P`, which makes for a lot of noise.

It would now be possible to combine the `reconcile()` and `maybeRollingUpdate()` operations for statefulsets, because we now always do them sequentially, but I've not done that in this PR because I thought it was already a big change.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [ ] ~Update documentation~
- [ ] ~Check RBAC rights for Kubernetes / OpenShift roles~
- [ ] ~Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally~
- [ ] ~Reference relevant issue(s) and close them after merging~

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java [hunk 1]
```diff
@@ -205,15 +208,53 @@ public void onClose(KubernetesClientException e) {
...
+            return -3;
+        }
+        return getGeneration(templateMetadata(resource));
+    }
+
+    private static int getPodGeneration(Pod resource) {
+        if (resource == null) {
+            return -4;
+        }
+        return getGeneration(resource.getMetadata());
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: Do these -3 and -4 values mean something useful?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java [hunk 2]
```diff
@@ -27,10 +28,12 @@
...
 
 /**
- * Operations for {@code StatefulSets}s, which supports {@link #rollingUpdate(String, String)}
+ * Operations for {@code StatefulSets}s, which supports {@link #maybeRollingUpdate(StatefulSet)}
  * in addition to the usual operations.
  */
-public class StatefulSetOperator<P> extends AbstractScalableResourceOperator<KubernetesClient, StatefulSet, StatefulSetList, DoneableStatefulSet, RollableScalableResource<StatefulSet, DoneableStatefulSet>, P> {
+public abstract class StatefulSetOperator extends AbstractScalableResourceOperator<KubernetesClient, StatefulSet, StatefulSetList, DoneableStatefulSet, RollableScalableResource<StatefulSet, DoneableStatefulSet>> {
+
+    public static final String ANNOTATION_GENERATION = "cluster-controller.strimzi.io/statefulset-generation";
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: why not using the `STRIMZI_DOMAIN` (just `strimzi.io`) we have in the `Labels` class?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java [hunk 3]
```diff
@@ -205,15 +208,53 @@ public void onClose(KubernetesClientException e) {
...
     }
 
     @Override
-    protected Future<ReconcileResult<P>> internalCreate(String namespace, String name, StatefulSet desired) {
+    protected Future<ReconcileResult<StatefulSet>> internalCreate(String namespace, String name, StatefulSet desired) {
         // Create the SS...
-        Future<ReconcileResult<P>> result = Future.future();
-        Future<ReconcileResult<P>> crt = super.internalCreate(namespace, name, desired);
+        Future<ReconcileResult<StatefulSet>> result = Future.future();
+        setGeneration(desired, 1);
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: Just a thought ... if we start from generation 0 on creation, the next generation values should represent the number of rolling updates so far. Maybe it's useful to know ... maybe not.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java (hunk 1 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: Ah thankyou, that's left over from debugging a weird problem I had with the tests, I'll fix this.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java (hunk 2 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: Actually shouldn't we use `ClusterOperator.STRIMZI_CLUSTER_OPERATOR_DOMAIN`, which is what we use for the delete claim annotation?

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: You are right but now I'm thinking ... does it make sense this specific cluster operator domain or the strimzi domain is enough. But up to you ... no strong opinion on that from me.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java (hunk 3 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: On the basis that annotations are more private impl detail than public API I think it's probably right that for those annotation which are impl detail purely of the CO, that they're scoped to the CO.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSetOperator.java [hunk 4]
```diff
@@ -29,24 +28,17 @@ public KafkaSetOperator(Vertx vertx, KubernetesClient client, long operationTime
...
-        } else {
-            boolean different = needsRollingUpdate(diff);
-            return super.internalPatch(namespace, name, current, desired).map(r -> {
-                if (r instanceof ReconcileResult.Patched) {
-                    return ReconcileResult.patched(different);
-                } else {
-                    return r;
-                }
-            });
+        if (diff.changesVolumeClaimTemplates()) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSetOperator.java]: Why do we have now this block twice here with one of them reffering to _Zookeeper_? Seems like a bug.

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSetOperator.java]: Good spot @scholzj

**Pull request merged**